src/qiskit_qasm3_import/expression.py

```diff
@@ -62,43 +76,60 @@ class ValueResolver(QASMVisitor):
     def __init__(self, symbols: Mapping[str, Symbol]):
         self.symbols = symbols
 
-    def resolve(self, node: ast.Expression) -> Tuple[Any, types.Type]:
+    def resolve(self, node: ast.Expression, context=None) -> Tuple[Any, types.Type]:
         """The entry point to the resolver, resolving the AST node into a 2-tuple of a relevant
         Qiskit type, and the :class:`.Type` that it is an instance of."""
-        return self.visit(node)
+        if context is None:
+            return self.visit(node)
+        return self.visit(node, context)
 
     def visit(self, node: ast.QASMNode, context: None = None) -> Tuple[Any, types.Type]:
-        value, type = super().visit(node, context)
+        if context is None:
+            value, type = super().visit(node)
+        else:
+            value, type = super().visit(node, context)
         if isinstance(type, types.Error):
             raise_from_node(node, "type error")
         return value, type
 
     def generic_visit(self, node: ast.QASMNode, context: None = None):
         raise_from_node(node, f"'{node.__class__.__name__}' cannot be resolved into a Qiskit value")
 
-    def visit_Identifier(self, node: ast.Identifier):
+    def visit_Identifier(self, node: ast.Identifier, context=None):
         name = node.name
         if name not in self.symbols:
-            raise_from_node(node, f"name '{name}' is not defined in this scope")
-        symbol = self.symbols[name]
+            if is_physical(name):  # Physical qubits are not declared.
+                if not context.addressing_mode.set_physical_mode():
+                    raise_from_node(
+                        node,
+                        "Physical qubit referenced in virtual addressing mode. Mixing modes not currently supported.",
+                    )
+                bit = Qubit()
+                symbol = Symbol(name, bit, types.Qubit(), Scope.GLOBAL, node)
+                context.circuit.add_bits([bit])
+                self.symbols[name] = symbol
+            else:
+                raise_from_node(node, f"name '{name}' is not defined in this scope")
```

Comment:
It's right to do the `is_physical` check here, but after that, we should be attaching it as information in the type system, rather than requiring other places to do the same semantic analysis later.